<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>


Before we get the data and start exploring it, let's import all the dependencies that we will need.


In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
from bs4 import BeautifulSoup # library to parse HTML

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 1. Download and Explore Dataset


I selected the table element from the wikipedia page and saved it in a separate file. The code below finds all elements in the table and saves the information.

In [2]:
fname = "toronto.html" # file with <table>
with open(fname) as fopen:
    data = fopen.read()

parse = BeautifulSoup(data, 'html.parser')
table = []
for td in parse.findAll('td'):
    if td.span.text == 'Not assigned':
        continue
    else:
        postcode = td.p.text[:3]
        borough = td.span.text.split('(')[0]
        neighborhoods = ' '.join(td.span.text.split('(')[1:]).strip(')').replace(')',' ').replace(' / ',',')
        if 'Not assigned' in neighborhoods:
            neighborhoods = borough
        #else:
        #    neighborhoods = neighborhoods.split(',')
        table.append({'postcode':postcode, 'borough':borough, 'neighborhoods':neighborhoods})
    
    
table = pd.DataFrame(table)
## Clean names of boroughs
table['borough']=table['borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

table
        

postcode                 borough  \
0        M3A              North York   
1        M4A              North York   
2        M5A        Downtown Toronto   
3        M6A              North York   
4        M7A            Queen's Park   
5        M9A               Etobicoke   
6        M1B             Scarborough   
7        M3B              North York   
8        M4B               East York   
9        M5B        Downtown Toronto   
10       M6B              North York   
11       M9B               Etobicoke   
12       M1C             Scarborough   
13       M3C              North York   
14       M4C               East York   
15       M5C        Downtown Toronto   
16       M6C                    York   
17       M9C               Etobicoke   
18       M1E             Scarborough   
19       M4E            East Toronto   
20       M5E        Downtown Toronto   
21       M6E                    York   
22       M1G             Scarborough   
23       M4G               East York   
24       M5G        Downtown Toronto   
25       M6G        Downtown Toronto   
26       M1H             Scarborough   
27       M2H              North York   
28       M3H              North York   
29       M4H               East York   
30       M5H        Downtown Toronto   
31       M6H            West Toronto   
32       M1J             Scarborough   
33       M2J              North York   
34       M3J              North York   
35       M4J  East York/East Toronto   
36       M5J        Downtown Toronto   
37       M6J            West Toronto   
38       M1K             Scarborough   
39       M2K              North York   
40       M3K              North York   
41       M4K            East Toronto   
42       M5K        Downtown Toronto   
43       M6K            West Toronto   
44       M1L             Scarborough   
45       M2L              North York   
46       M3L              North York   
47       M4L            East Toronto   
48       M5L        Downtown Toronto   
49       M6L              North York   
50       M9L              North York   
51       M1M             Scarborough   
52       M2M              North York   
53       M3M              North York   
54       M4M            East Toronto   
55       M5M              North York   
56       M6M                    York   
57       M9M              North York   
58       M1N             Scarborough   
59       M2N              North York   
60       M3N              North York   
61       M4N         Central Toronto   
62       M5N         Central Toronto   
63       M6N                    York   
64       M9N                    York   
65       M1P             Scarborough   
66       M2P              North York   
67       M4P         Central Toronto   
68       M5P         Central Toronto   
69       M6P            West Toronto   
70       M9P               Etobicoke   
71       M1R             Scarborough   
72       M2R              North York   
73       M4R         Central Toronto   
74       M5R         Central Toronto   
75       M6R            West Toronto   
76       M7R             Mississauga   
77       M9R               Etobicoke   
78       M1S             Scarborough   
79       M4S         Central Toronto   
80       M5S        Downtown Toronto   
81       M6S            West Toronto   
82       M1T             Scarborough   
83       M4T         Central Toronto   
84       M5T        Downtown Toronto   
85       M1V             Scarborough   
86       M4V         Central Toronto   
87       M5V        Downtown Toronto   
88       M8V               Etobicoke   
89       M9V               Etobicoke   
90       M1W             Scarborough   
91       M4W        Downtown Toronto   
92       M5W  Downtown Toronto Stn A   
93       M8W               Etobicoke   
94       M9W     Etobicoke Northwest   
95       M1X             Scarborough   
96       M4X        Downtown Toronto   
97       M5X        Downtown Toronto   
98       M8X               Etobicoke   
99  

#### Load and explore the data


Next, let's load the data.


In [3]:
fname = 'Geospatial_Coordinates.csv'
## Align both data frames so rows are consistent and can be merged
table = table.sort_values('postcode').reset_index().drop('index',axis=1)
coords_data = pd.read_csv(fname).sort_values('Postal Code').reset_index().drop('index',axis=1)
#print(table.tail())
#print(coords_data.tail())
table['latitude'] = coords_data['Latitude']
table['longitude'] = coords_data['Longitude']
table.shape

(103, 5)

Quickly examine the resulting dataframe.


In [4]:
table.head()

postcode      borough                         neighborhoods   latitude  \
0      M1B  Scarborough                         Malvern,Rouge  43.806686   
1      M1C  Scarborough  Rouge Hill,Port Union,Highland Creek  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

And make sure that the dataset has all 15 boroughs and 103 postal codes.


In [5]:
print('The dataframe has {} boroughs and {} postal codes.'.format(
        len(table['borough'].unique()),
        table.shape[0]
    )
)

The dataframe has 15 boroughs and 103 postal codes.


#### Use geopy library to get the latitude and longitude values of Toronto.


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>toronto_explorer</em>, as shown below.


In [6]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with postal codes superimposed on top.


In [7]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, postcode in zip(table['latitude'], table['longitude'], table['borough'], table['postcode']):
    label = '{}, {}'.format(postcode, borough)
    color = 'yellow' # M1*
    if '2' in postcode: color = 'red' # M2
    elif '3' in postcode: color = 'blue' # M3
    elif '4' in postcode: color = 'green' # M4
    elif '5' in postcode: color = 'orange' # M5
    elif '6' in postcode: color = 'purple' # M6
    elif '7' in postcode: color = 'cyan' # M7
    elif '8' in postcode: color = 'gray' # M8
    elif '9' in postcode: color = 'white' # M9
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color=color,
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

However, for illustration purposes, let's simplify the above map and segment and cluster only the post codes starting with M5. So let's slice the original dataframe and create a new dataframe of the Toronto data.


In [8]:
m5_data = table[table['postcode'].str.startswith('M5')].reset_index(drop=True)
m5_data.head()

postcode           borough             neighborhoods   latitude  longitude
0      M5A  Downtown Toronto  Regent Park,Harbourfront  43.654260 -79.360636
1      M5B  Downtown Toronto  Garden District, Ryerson  43.657162 -79.378937
2      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418
3      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306
4      M5G  Downtown Toronto        Central Bay Street  43.657952 -79.387383

Let's get the geographical coordinates of Downtown Toronto.


In [9]:
address = 'Downtown Toronto, Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Downtown Toronto are 43.6563221, -79.3809161.


As we did with all of Toronto, let's visualize Downtown Toronto and the post codes in it.


In [10]:
# create map of Downtown Toronto using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(m5_data['latitude'], m5_data['longitude'], m5_data['postcode']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

Next, we are going to start utilizing the Foursquare API to explore the post codes and segment them.


#### Define Foursquare Credentials and Version


In [11]:
with open('.fs_client_id') as fopen:
    CLIENT_ID = fopen.read().strip() # your Foursquare ID
with open('.fs_client_secret') as fopen:
    CLIENT_SECRET = fopen.read().strip() # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:


#### Let's explore the first postcode in our dataframe.


Get the postcode's name.


In [12]:
m5_data.loc[0, 'postcode']

'M5A'

Get the postcode's latitude and longitude values.


In [13]:
postcode_latitude = m5_data.loc[0, 'latitude'] # neighborhood latitude value
postcode_longitude = m5_data.loc[0, 'longitude'] # neighborhood longitude value

postcode_name = m5_data.loc[0, 'postcode'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(postcode_name, 
                                                               postcode_latitude, 
                                                               postcode_longitude))

Latitude and longitude values of M5A are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in M5A within a radius of 500 meters.


First, let's create the GET request URL. Name your URL **url**.


In [14]:
# type your answer here
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, postcode_latitude, postcode_longitude, VERSION, radius, LIMIT) 
url


'https://api.foursquare.com/v2/venues/explore?client_id=45G3AFA1RMWI5ZBGA1CNMWIANV2IKBMQZ2055Q3FY4UFUWZ0&client_secret=CXGDLJOHSB3LKYZ2RIGELH5C2RS4L2CLRJHDWYLY4F1KO32S&ll=43.6542599,-79.3606359&v=20180605&radius=500&limit=100'

Double-click **here** for the solution.

<!-- The correct answer is:
LIMIT = 100 # limit of number of venues returned by Foursquare API
-->

<!--
radius = 500 # define radius
-->

<!--
\\\\ # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL
--> 


Send the GET request and examine the resutls


In [15]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '61088edef2332049b48270df'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 45,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

From the Foursquare lab in the previous module, we know that all the information is in the _items_ key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.


In [16]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a _pandas_ dataframe.


In [17]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

And how many venues were returned by Foursquare?


In [18]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

45 venues were returned by Foursquare.


## 2. Explore Postcodes starting with M5


#### Let's create a function to repeat the same process to all the postcodes that start with M5


In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postcode', 
                  'Postcode Latitude', 
                  'Postcode Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called _manhattan_venues_.


In [20]:
# type your answer here
m5_venues = getNearbyVenues( m5_data['postcode'], m5_data['latitude'], m5_data['longitude'], radius = 1000)
m5_venues

M5A
M5B
M5C
M5E
M5G
M5H
M5J
M5K
M5L
M5M
M5N
M5P
M5R
M5S
M5T
M5V
M5W
M5X


Postcode  Postcode Latitude  Postcode Longitude  \
0         M5A          43.654260          -79.360636   
1         M5A          43.654260          -79.360636   
2         M5A          43.654260          -79.360636   
3         M5A          43.654260          -79.360636   
4         M5A          43.654260          -79.360636   
5         M5A          43.654260          -79.360636   
6         M5A          43.654260          -79.360636   
7         M5A          43.654260          -79.360636   
8         M5A          43.654260          -79.360636   
9         M5A          43.654260          -79.360636   
10        M5A          43.654260          -79.360636   
11        M5A          43.654260          -79.360636   
12        M5A          43.654260          -79.360636   
13        M5A          43.654260          -79.360636   
14        M5A          43.654260          -79.360636   
15        M5A          43.654260          -79.360636   
16        M5A          43.654260          -79.360636   
17        M5A          43.654260          -79.360636   
18        M5A          43.654260          -79.360636   
19        M5A          43.654260          -79.360636   
20        M5A          43.654260          -79.360636   
21        M5A          43.654260          -79.360636   
22        M5A          43.654260          -79.360636   
23        M5A          43.654260          -79.360636   
24        M5A          43.654260          -79.360636   
25        M5A          43.654260          -79.360636   
26        M5A          43.654260          -79.360636   
27        M5A          43.654260          -79.360636   
28        M5A          43.654260          -79.360636   
29        M5A          43.654260          -79.360636   
30        M5A          43.654260          -79.360636   
31        M5A          43.654260          -79.360636   
32        M5A          43.654260          -79.360636   
33        M5A          43.654260          -79.360636   
34        M5A          43.654260          -79.360636   
35        M5A          43.654260          -79.360636   
36        M5A          43.654260          -79.360636   
37        M5A          43.654260          -79.360636   
38        M5A          43.654260          -79.360636   
39        M5A          43.654260          -79.360636   
40        M5A          43.654260          -79.360636   
41        M5A          43.654260          -79.360636   
42        M5A          43.654260          -79.360636   
43        M5A          43.654260          -79.360636   
44        M5A          43.654260          -79.360636   
45        M5A          43.654260          -79.360636   
46        M5A          43.654260          -79.360636   
47        M5A          43.654260          -79.360636   
48        M5A          43.654260          -79.360636   
49        M5A          43.654260          -79.360636   
50        M5A          43.654260          -79.360636   
51        M5A          43.654260          -79.360636   
52        M5A          43.654260          -79.360636   
53        M5A          43.654260          -79.360636   
54        M5A          43.654260          -79.360636   
55        M5A          43.654260          -79.360636   
56        M5A          43.654260          -79.360636   
57        M5A          43.654260          -79.360636   
58        M5A          43.654260          -79.360636   
59        M5A          43.654260          -79.360636   
60        M5A          43.654260          -79.360636   
61        M5A          43.654260          -79.360636   
62        M5A          43.654260          -79.360636   
63        M5A          43.654260          -79.360636   
64        M5A          43.654260          -79.360636   
65        M5A          43.654260          -79.360636   
66        M5A          43.654260          -79.360636   
67        M5A          43.654260          -79.360636   
68        M5A          43.654260          -79.360636   
69        M5A          43.654260          -79.360636   
70        M5A          43.654

Double-click **here** for the solution.

<!-- The correct answer is:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )
--> 


#### Let's check the size of the resulting dataframe


In [21]:
print(m5_venues.shape)
m5_venues.head()

(1529, 7)


Postcode  Postcode Latitude  Postcode Longitude  \
0      M5A           43.65426          -79.360636   
1      M5A           43.65426          -79.360636   
2      M5A           43.65426          -79.360636   
3      M5A           43.65426          -79.360636   
4      M5A           43.65426          -79.360636   

                              Venue  Venue Latitude  Venue Longitude  \
0                  Roselle Desserts       43.653447       -79.362017   
1                     Tandem Coffee       43.653559       -79.361809   
2            Cooper Koo Family YMCA       43.653249       -79.358008   
3                    Impact Kitchen       43.656369       -79.356980   
4  The Distillery Historic District       43.650244       -79.359323   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4        Historic Site

Let's check how many venues were returned for each neighborhood


In [22]:
m5_venues.groupby('Postcode').count()

Postcode Latitude  Postcode Longitude  Venue  Venue Latitude  \
Postcode                                                                 
M5A                     100                 100    100             100   
M5B                     100                 100    100             100   
M5C                     100                 100    100             100   
M5E                     100                 100    100             100   
M5G                     100                 100    100             100   
M5H                     100                 100    100             100   
M5J                     100                 100    100             100   
M5K                     100                 100    100             100   
M5L                     100                 100    100             100   
M5M                      41                  41     41              41   
M5N                      27                  27     27              27   
M5P                      46                  46     46              46   
M5R                     100                 100    100             100   
M5S                     100                 100    100             100   
M5T                     100                 100    100             100   
M5V                      15                  15     15              15   
M5W                     100                 100    100             100   
M5X                     100                 100    100             100   

          Venue Longitude  Venue Category  
Postcode                                   
M5A                   100             100  
M5B                   100             100  
M5C                   100             100  
M5E                   100             100  
M5G                   100             100  
M5H                   100             100  
M5J                   100             100  
M5K                   100             100  
M5L                   100             100  
M5M                    41              41  
M5N                    27              27  
M5P                    46              46  
M5R                   100             100  
M5S                   100             100  
M5T                   100             100  
M5V                    15              15  
M5W                   100             100  
M5X                   100             100

#### Let's find out how many unique categories can be curated from all the returned venues


In [23]:
print('There are {} uniques categories.'.format(len(m5_venues['Venue Category'].unique())))

There are 197 uniques categories.


## 3. Analyze Each Neighborhood


In [24]:
# one hot encoding
m5_onehot = pd.get_dummies(m5_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
m5_onehot['Postcode'] = m5_venues['Postcode'] 

# move neighborhood column to the first column
fixed_columns = [m5_onehot.columns[-1]] + list(m5_onehot.columns[:-1])
m5_onehot = m5_onehot[fixed_columns]

m5_onehot.head()

Postcode  Airport  Airport Lounge  American Restaurant  Animal Shelter  \
0      M5A        0               0                    0               0   
1      M5A        0               0                    0               0   
2      M5A        0               0                    0               0   
3      M5A        0               0                    0               0   
4      M5A        0               0                    0               0   

   Aquarium  Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
0         0       0            0           0                    0   
1         0       0            0           0                    0   
2         0       0            0           0                    0   
3         0       0            0           0                    0   
4         0       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  BBQ Joint  \
0                 0                   0                0          0   
1                 0                   0                0          0   
2                 0                   0                0          0   
3                 0                   0                0          0   
4                 0                   0                0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       1     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bike Shop  Bistro  Bookstore  Botanical Garden  Boutique  \
0          0       0          0                 0         0   
1          0       0          0                 0         0   
2          0       0          0                 0         0   
3          0       0          0                 0         0   
4          0       0          0                 0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Butcher  Café  Camera Store  \
0                0             0              0        0     0             0   
1                0             0              0        0     0             0   
2                0             0              0        0     0             0   
3                0             0              0        0     0             0   
4                0             0              0        0     0             0   

   Caribbean Restaurant  Castle  Cheese Shop  Chinese Restaurant  \
0                     0       0            0                   0   
1                     0       0            0                   0   
2                     0       0            0                   0   
3                     0       0            0                   0   
4                     0       0            0                   0   

   Chocolate Shop  Clothing Store  Cocktail Bar  Coffee Shop  College Gym  \
0               0               0             0            0            0   
1               0               0             0

And let's examine the new dataframe size.


In [25]:
m5_onehot.shape

(1529, 198)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [26]:
m5_grouped = m5_onehot.groupby('Postcode').mean().reset_index()
m5_grouped

Postcode   Airport  Airport Lounge  American Restaurant  Animal Shelter  \
0       M5A  0.000000        0.000000              0.00000            0.01   
1       M5B  0.000000        0.000000              0.01000            0.00   
2       M5C  0.000000        0.000000              0.02000            0.00   
3       M5E  0.000000        0.000000              0.02000            0.00   
4       M5G  0.000000        0.000000              0.01000            0.00   
5       M5H  0.000000        0.000000              0.02000            0.00   
6       M5J  0.000000        0.000000              0.00000            0.00   
7       M5K  0.000000        0.000000              0.02000            0.00   
8       M5L  0.000000        0.000000              0.02000            0.00   
9       M5M  0.000000        0.000000              0.02439            0.00   
10      M5N  0.000000        0.000000              0.00000            0.00   
11      M5P  0.000000        0.000000              0.00000            0.00   
12      M5R  0.000000        0.000000              0.01000            0.00   
13      M5S  0.000000        0.000000              0.00000            0.00   
14      M5T  0.000000        0.000000              0.00000            0.00   
15      M5V  0.066667        0.066667              0.00000            0.00   
16      M5W  0.000000        0.000000              0.02000            0.00   
17      M5X  0.000000        0.000000              0.02000            0.00   

    Aquarium  Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
0       0.00    0.00         0.01        0.00                 0.00   
1       0.00    0.00         0.00        0.01                 0.00   
2       0.00    0.00         0.01        0.00                 0.00   
3       0.00    0.00         0.02        0.00                 0.00   
4       0.00    0.00         0.01        0.01                 0.01   
5       0.00    0.00         0.02        0.01                 0.02   
6       0.01    0.00         0.01        0.00                 0.00   
7       0.01    0.00         0.01        0.00                 0.00   
8       0.00    0.00         0.02        0.00                 0.00   
9       0.00    0.00         0.00        0.00                 0.00   
10      0.00    0.00         0.00        0.00                 0.00   
11      0.00    0.00         0.00        0.00                 0.00   
12      0.00    0.00         0.00        0.00                 0.01   
13      0.00    0.00         0.00        0.00                 0.02   
14      0.00    0.01         0.03        0.00                 0.02   
15      0.00    0.00         0.00        0.00                 0.00   
16      0.00    0.00         0.02        0.00                 0.00   
17      0.00    0.00         0.01        0.00                 0.00   

    Asian Restaurant  Athletics & Sports  Auto Dealership  BBQ Joint  \
0           0.010000                0.01             0.01       0.00   
1           0.000000                0.00             0.00       0.01   
2           0.000000                0.00             0.00       0.01   
3           0.000000                0.00             0.00       0.01   
4           0.000000                0.00             0.00       0.00   
5           0.010000                0.00             0.00       0.00   
6           0.000000                0.00             0.00       0.00   
7           0.010000                0.00             0.00       0.00   
8           0.010000                0.00             0.00       0.00   
9           0.000000                0.00             0.00       0.00   
10          0.000000                0.00             0.00       0.00   
11          0.021739                0.00             0.00       0.00   
12          0.000000                0.00             0.00       0.01   
13          0.000000                0.00             0.00       0.00   
14          0.000000                0.00             0.00       0.00   
15          0.000000                0

#### Let's confirm the new size


In [27]:
m5_grouped.shape

(18, 198)

#### Let's print each neighborhood along with the top 5 most common venues


In [28]:
num_top_venues = 5

for hood in m5_grouped['Postcode']:
    hoods = table[table['postcode'] == hood].values[0][2]
    print("----"+hoods+"----")
    temp = m5_grouped[m5_grouped['Postcode'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 3})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Regent Park,Harbourfront----
         venue  freq
0  Coffee Shop  0.15
1         Café  0.04
2      Theater  0.04
3       Bakery  0.04
4          Pub  0.04


----Garden District, Ryerson----
                venue  freq
0         Coffee Shop  0.09
1      Clothing Store  0.05
2           Gastropub  0.04
3  Italian Restaurant  0.04
4      Cosmetics Shop  0.03


----St. James Town----
            venue  freq
0            Café  0.08
1     Coffee Shop  0.08
2      Restaurant  0.06
3          Bakery  0.04
4  Clothing Store  0.03


----Berczy Park----
                 venue  freq
0          Coffee Shop  0.09
1                 Café  0.06
2           Restaurant  0.06
3                Hotel  0.05
4  Japanese Restaurant  0.04


----Central Bay Street----
              venue  freq
0       Coffee Shop  0.09
1              Café  0.04
2  Sushi Restaurant  0.04
3              Park  0.03
4         Gastropub  0.03


----Richmond,Adelaide,King----
                 venue  freq
0          Coffee Shop  0.

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.


In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postcodes_venues_sorted = pd.DataFrame(columns=columns)
postcodes_venues_sorted['Postcode'] = m5_grouped['Postcode']

for ind in np.arange(m5_grouped.shape[0]):
    postcodes_venues_sorted.iloc[ind, 1:] = return_most_common_venues(m5_grouped.iloc[ind, :], num_top_venues)

postcodes_venues_sorted.head()


Postcode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0      M5A           Coffee Shop                   Pub                  Café   
1      M5B           Coffee Shop        Clothing Store    Italian Restaurant   
2      M5C           Coffee Shop                  Café            Restaurant   
3      M5E           Coffee Shop            Restaurant                  Café   
4      M5G           Coffee Shop      Sushi Restaurant                  Café   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Bakery                  Park               Theater   
1             Gastropub        Cosmetics Shop   Japanese Restaurant   
2                Bakery    Italian Restaurant    Seafood Restaurant   
3                 Hotel   Japanese Restaurant                  Park   
4                  Park             Gastropub        Clothing Store   

  7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
0                 Diner            Restaurant           Sushi Restaurant   
1            Restaurant                 Diner  Middle Eastern Restaurant   
2                 Hotel        Clothing Store        Japanese Restaurant   
3                Bakery              Beer Bar               Cocktail Bar   
4           Yoga Studio      Ramen Restaurant         Mexican Restaurant   

  10th Most Common Venue  
0         Breakfast Spot  
1          Burrito Place  
2              Gastropub  
3                    Gym  
4    Japanese Restaurant

## 4. Cluster Postcodes


Run _k_-means to cluster the postcodes into 3 clusters.


In [31]:
# set number of clusters
kclusters = 3

m5_grouped_clustering = m5_grouped.drop('Postcode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(m5_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 2, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each postcode.


In [32]:
# add clustering labels
postcodes_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

m5_merged = m5_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
m5_merged = m5_merged.join(postcodes_venues_sorted.set_index('Postcode'), on='postcode')

m5_merged.head() # check the last columns!

postcode           borough             neighborhoods   latitude  longitude  \
0      M5A  Downtown Toronto  Regent Park,Harbourfront  43.654260 -79.360636   
1      M5B  Downtown Toronto  Garden District, Ryerson  43.657162 -79.378937   
2      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418   
3      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306   
4      M5G  Downtown Toronto        Central Bay Street  43.657952 -79.387383   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1           Coffee Shop                   Pub   
1               1           Coffee Shop        Clothing Store   
2               1           Coffee Shop                  Café   
3               1           Coffee Shop            Restaurant   
4               1           Coffee Shop      Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Café                Bakery                  Park   
1    Italian Restaurant             Gastropub        Cosmetics Shop   
2            Restaurant                Bakery    Italian Restaurant   
3                  Café                 Hotel   Japanese Restaurant   
4                  Café                  Park             Gastropub   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Theater                 Diner            Restaurant   
1   Japanese Restaurant            Restaurant                 Diner   
2    Seafood Restaurant                 Hotel        Clothing Store   
3                  Park                Bakery              Beer Bar   
4        Clothing Store           Yoga Studio      Ramen Restaurant   

       9th Most Common Venue 10th Most Common Venue  
0           Sushi Restaurant         Breakfast Spot  
1  Middle Eastern Restaurant          Burrito Place  
2        Japanese Restaurant              Gastropub  
3               Cocktail Bar                    Gym  
4         Mexican Restaurant    Japanese Restaurant

Finally, let's visualize the resulting clusters


In [35]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(m5_merged['latitude'], m5_merged['longitude'], m5_merged['postcode'], m5_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>


## 5. Examine Clusters


Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.


#### Cluster 0


In [36]:
m5_merged.loc[m5_merged['Cluster Labels'] == 0, m5_merged.columns[[1] + list(range(5, m5_merged.shape[1]))]]

borough  Cluster Labels 1st Most Common Venue  \
9        North York               0           Coffee Shop   
10  Central Toronto               0      Sushi Restaurant   
11  Central Toronto               0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9     Italian Restaurant                  Bank  Fast Food Restaurant   
10    Italian Restaurant                   Spa              Pharmacy   
11                  Café                  Bank           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9         Sandwich Place           Bridal Shop     Indian Restaurant   
10           Coffee Shop                  Café                  Bank   
11      Sushi Restaurant          Liquor Store   Japanese Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9           Intersection          Skating Rink               Boutique  
10            Playground          Skating Rink            Dry Cleaner  
11    Italian Restaurant            Playground               Pharmacy

#### Cluster 1


In [37]:
m5_merged.loc[m5_merged['Cluster Labels'] == 1, m5_merged.columns[[1] + list(range(5, m5_merged.shape[1]))]]

borough  Cluster Labels 1st Most Common Venue  \
0         Downtown Toronto               1           Coffee Shop   
1         Downtown Toronto               1           Coffee Shop   
2         Downtown Toronto               1           Coffee Shop   
3         Downtown Toronto               1           Coffee Shop   
4         Downtown Toronto               1           Coffee Shop   
5         Downtown Toronto               1           Coffee Shop   
6         Downtown Toronto               1                  Café   
7         Downtown Toronto               1           Coffee Shop   
8         Downtown Toronto               1           Coffee Shop   
12         Central Toronto               1    Italian Restaurant   
13        Downtown Toronto               1                  Café   
14        Downtown Toronto               1                  Café   
16  Downtown Toronto Stn A               1           Coffee Shop   
17        Downtown Toronto               1           Coffee Shop   

   2nd Most Common Venue          3rd Most Common Venue  \
0                    Pub                           Café   
1         Clothing Store             Italian Restaurant   
2                   Café                     Restaurant   
3             Restaurant                           Café   
4       Sushi Restaurant                           Café   
5                   Café                          Hotel   
6            Coffee Shop                          Hotel   
7                   Café                          Hotel   
8                  Hotel                     Restaurant   
12                  Café                    Coffee Shop   
13                   Bar                         Bakery   
14                   Bar  Vegetarian / Vegan Restaurant   
16            Restaurant                           Café   
17                 Hotel                           Café   

            4th Most Common Venue 5th Most Common Venue  \
0                          Bakery                  Park   
1                       Gastropub        Cosmetics Shop   
2                          Bakery    Italian Restaurant   
3                           Hotel   Japanese Restaurant   
4                            Park             Gastropub   
5                         Theater            Restaurant   
6                            Park   Japanese Restaurant   
7                      Restaurant   Japanese Restaurant   
8                            Café                Bakery   
12                            Gym            Restaurant   
13  Vegetarian / Vegan Restaurant             Bookstore   
14                    Coffee Shop           Art Gallery   
16                         Bakery   Japanese Restaurant   
17                     Restaurant               Theater   

            6th Most Common Venue 7th Most Common Venue  \
0                         Theater                 Diner   
1             Japanese Restaurant            Restaurant   
2              Seafood Restaurant                 Hotel   
3                            Park                Bakery   
4                  Clothing Store           Yoga Studio   
5             Japanese Restaurant        Clothing Store   
6                      Restaurant                   Gym   
7                         Theater                Bakery   
8              Seafood Restaurant        Cosmetics Shop   
12             Mexican Restaurant         Grocery Store   
13                    Coffee Shop            Restaurant   
14           Caribbean Restaurant           Yoga Studio   
16                          Hotel                  Park   
17  Vegetarian / Vegan Restaurant   Japanese Restaurant   

            8th Most Common Venue      9th Most Common Venue  \
0                      Restaurant           Sushi Restaurant   
1                           Diner  Middle Eastern Restaurant   
2                  Clothing Store        Japanese Restaurant   
3                        Beer Bar               Cocktail Bar   
4                Ram

#### Cluster 2


In [38]:
m5_merged.loc[m5_merged['Cluster Labels'] == 2, m5_merged.columns[[1] + list(range(5, m5_merged.shape[1]))]]

borough  Cluster Labels 1st Most Common Venue  \
15  Downtown Toronto               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
15                  Café       Harbor / Marina               Airport   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
15                 Track      Sculpture Garden        Scenic Lookout   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
15               Dog Run                Garden           Dance Studio